tests/federation_room_get_missing_events_test.go

```diff
@@ -314,3 +313,135 @@ func TestOutboundFederationIgnoresMissingEventWithBadJSONForRoomVersion6(t *test
+	defer cancel()
+
+	charlie := srv.UserID("charlie")
+	roomVersion := alice.GetDefaultRoomVersion(t)
+
+	for _, visibility := range []gomatrixserverlib.HistoryVisibility{
+		gomatrixserverlib.HistoryVisibilityWorldReadable,
+		gomatrixserverlib.HistoryVisibilityShared,
+		gomatrixserverlib.HistoryVisibilityInvited,
+		gomatrixserverlib.HistoryVisibilityJoined,
+	} {
+		// sytest: Inbound federation can return missing events for $vis visibility
+		t.Run(fmt.Sprintf("Inbound federation can return missing events for %s visibility", visibility), func(t *testing.T) {
+			roomID := alice.CreateRoom(t, map[string]interface{}{
+				"preset":  "public_chat",
+				"version": roomVersion,
+			})
+
+			stateKey := ""
+			// Set the history visibility for this room
+			alice.SendEventSynced(t, roomID, b.Event{
+				Type: gomatrixserverlib.MRoomHistoryVisibility,
+				Content: map[string]interface{}{
+					"history_visibility": visibility,
+				},
+				StateKey: &stateKey,
+			})
+
+			alice.SendEventSynced(t, roomID, b.Event{
+				Type: "m.room.message",
+				Content: map[string]interface{}{
+					"body":    "1",
+					"msgtype": "m.text",
+				},
+				Sender: alice.UserID,
+			})
+
+			room := srv.MustJoinRoom(t, deployment, "hs1", roomID, charlie)
+			alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(charlie, roomID))
+
+			req := gomatrixserverlib.NewFederationRequest("POST", "hs1",
+				fmt.Sprintf("/_matrix/federation/v1/get_missing_events/%s", roomID),
+			)
+
+			// Find two event IDs that there's going to be something missing
+			// inbetween. Say, any history between the room's creation and my own
+			// joining of it.
+			earliestEvent := room.CurrentState("m.room.create", "")
+			latestEvent := room.CurrentState("m.room.member", charlie)
+
+			err := req.SetContent(map[string]interface{}{
+				"earliest_events": []string{earliestEvent.EventID()},
+				"latest_events":   []string{latestEvent.EventID()},
+				"limit":           10,
+				"min_depth":       1,
+			})
+			must.NotError(t, "failed to set content", err)
+
+			result := gomatrixserverlib.RespMissingEvents{}
+			err = srv.SendFederationRequest(context.Background(), t, deployment, req, &result)
+			must.NotError(t, "get_missing_events failed", err)
+			if len(result.Events) == 0 {
+				t.Fatalf("no events returned from /get_missing_events")
+			}
+
+			// check that they are the *right* events. We expect copies of:
+			// * the creator's join
+			// * the power_levels
+			// * the join rules
+			// * the initial history_vis
+			// * another history_vis, unless we tried to set it to the default (shared)
+			// * the message
+			events := result.Events.UntrustedEvents(roomVersion)
+			i := 0
+			for _, ev := range events {
+				must.EqualStr(t, ev.RoomID(), roomID, "unexpected roomID")
+				switch i {
+				case 0:
+					must.EqualStr(t, ev.Type(), "m.room.member", "not a membership event")
+					must.EqualStr(t, *ev.StateKey(), alice.UserID, "unexpected creator")
+					i++
```

Comment:
Would it be easier to loop over `i, ev := range events` here and additionally check that `len(events)` is the correct size?

Reply:
I've updated the test. Now we're also checking that we received the message, if the visibility is `shared`